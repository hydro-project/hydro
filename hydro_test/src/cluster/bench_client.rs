use std::cell::RefCell;
use std::rc::Rc;
use std::time::Duration;

use hydro_lang::*;
use stats_ci::{Confidence, StatisticsOps};
use tokio::time::Instant;

pub struct Client {}

pub fn bench_client<'a>(
    clients: &Cluster<'a, Client>,
    transaction_cycle: impl FnOnce(
        Stream<(u32, u32), Cluster<'a, Client>, Unbounded>,
    ) -> Stream<(u32, u32), Cluster<'a, Client>, Unbounded, NoOrder>,
    num_clients_per_node: usize,
    median_latency_window_size: usize,
) {
    let client_tick = clients.tick();
    // r_to_clients_payload_applied.clone().inspect(q!(|payload: &(u32, ReplicaPayload)| println!("Client received payload: {:?}", payload)));

    // Set up an initial set of payloads on the first tick
    let start_this_tick = client_tick.optional_first_tick(q!(()));

    let c_new_payloads_on_start = start_this_tick.clone().flat_map_ordered(q!(move |_| (0
        ..num_clients_per_node)
        .map(move |i| (
            (CLUSTER_SELF_ID.raw_id * (num_clients_per_node as u32)) + i as u32,
            0
        ))));

    let (c_to_proposers_complete_cycle, c_to_proposers) =
        clients.forward_ref::<Stream<_, _, _, TotalOrder>>();
    let c_received_quorum_payloads = unsafe {
        // SAFETY: because the transaction processor is required to handle arbitrary reordering
        // across *different* keys, we are safe because delaying a transaction result for a key
        // will only affect when the next request for that key is emitted with respect to other
        // keys
        transaction_cycle(c_to_proposers).tick_batch(&client_tick)
    };

    // Whenever all replicas confirm that a payload was committed, send another payload
    let c_new_payloads_when_committed = c_received_quorum_payloads
        .clone()
        .map(q!(|payload| (payload.0, payload.1 + 1)));
    c_to_proposers_complete_cycle.complete(
        c_new_payloads_on_start
            .chain(unsafe {
                // SAFETY: we don't send a new write for the same key until the previous one is committed,
                // so this contains only a single write per key, and we don't care about order
                // across keys
                c_new_payloads_when_committed.assume_ordering::<TotalOrder>()
            })
            .all_ticks(),
    );

    // Track statistics
    let (c_timers_complete_cycle, c_timers) =
        client_tick.cycle::<Stream<(usize, Instant), _, _, NoOrder>>();
    let c_new_timers_when_leader_elected = start_this_tick
        .map(q!(|_| Instant::now()))
        .flat_map_ordered(q!(
            move |now| (0..num_clients_per_node).map(move |virtual_id| (virtual_id, now))
        ));
    let c_updated_timers = c_received_quorum_payloads
        .clone()
        .map(q!(|(key, _prev_count)| (key as usize, Instant::now())));
    let c_new_timers = c_timers
        .clone() // Update c_timers in tick+1 so we can record differences during this tick (to track latency)
        .chain(c_new_timers_when_leader_elected)
        .chain(c_updated_timers.clone())
        .reduce_keyed_commutative(q!(|curr_time, new_time| {
            if new_time > *curr_time {
                *curr_time = new_time;
            }
        }));
    c_timers_complete_cycle.complete_next_tick(c_new_timers);

    let c_stats_output_timer = unsafe {
        // SAFETY: intentionally sampling statistics
        clients
            .source_interval(q!(Duration::from_secs(1)))
            .tick_batch(&client_tick)
    }
    .first();

    let c_latency_reset = c_stats_output_timer.clone().map(q!(|_| None)).defer_tick();

    let c_latencies = c_timers
        .join(c_updated_timers)
        .map(q!(|(_virtual_id, (prev_time, curr_time))| Some(
            curr_time.duration_since(prev_time)
        )))
        .chain(c_latency_reset.into_stream())
        .all_ticks()
        .flatten_ordered()
        .fold_commutative(
            // Create window with ring buffer using vec + wraparound index
            // TODO: Would be nice if I could use vec![] instead, but that doesn't work in Hydro with RuntimeData *median_latency_window_size
            q!(move || (
                Rc::new(RefCell::new(Vec::<Duration>::with_capacity(
                    median_latency_window_size
                ))),
                0usize,
            )),
            q!(move |(latencies, write_index), latency| {
                let mut latencies_mut = latencies.borrow_mut();
                if *write_index < latencies_mut.len() {
                    latencies_mut[*write_index] = latency;
                } else {
                    latencies_mut.push(latency);
                }
                // Increment write index and wrap around
                *write_index = (*write_index + 1) % median_latency_window_size;
            }),
        )
        .map(q!(|(latencies, _)| latencies));

    let c_throughput_new_batch = c_received_quorum_payloads
        .clone()
        .count()
        .continue_unless(c_stats_output_timer.clone())
        .map(q!(|batch_size| (batch_size, false)));

    let c_throughput_reset = c_stats_output_timer
        .clone()
        .map(q!(|_| (0, true)))
        .defer_tick();

    let c_throughput = c_throughput_new_batch
        .union(c_throughput_reset)
        .all_ticks()
        .fold(
            q!(|| (0, { stats_ci::mean::Arithmetic::new() })),
            q!(|(total, stats), (batch_size, reset)| {
                if reset {
                    if *total > 0 {
                        stats.extend(&[*total as f64]).unwrap();
                    }

                    *total = 0;
                } else {
                    *total += batch_size;
                }
            }),
        )
        .map(q!(|(_, stats)| { stats }));

    unsafe {
        // SAFETY: intentionally sampling statistics
        c_latencies
            .latest_tick(&client_tick)
            .zip(c_throughput.latest_tick(&client_tick))
    }
    .continue_if(c_stats_output_timer)
    .all_ticks()
    .for_each(q!(move |(latencies, throughput)| {
        let mut latencies_mut = latencies.borrow_mut();

        let confidence = Confidence::new(0.95);

        latencies_mut.sort_unstable();
        if let Ok(interval) =
            stats_ci::quantile::ci_sorted_unchecked(confidence, latencies_mut.as_slice(), 0.5)
        {
            if let Some(lower) = interval.left() {
                if let Some(upper) = interval.right() {
                    println!(
                        "Latency Median 95% interval: {:.2} - {:.2} ms",
                        lower.as_micros() as f64 / 1000.0,
                        upper.as_micros() as f64 / 1000.0
                    );
                }
            }
        }

        if throughput.sample_count() >= 2 {
            // ci_mean crashes if there are fewer than two samples
            if let Ok(interval) = throughput.ci_mean(confidence) {
                if let Some(lower) = interval.left() {
                    if let Some(upper) = interval.right() {
                        println!(
                            "Throughput 95% interval: {:.2} - {:.2} requests/s",
                            lower, upper
                        );
                    }
                }
            }
        }
    }));
    // End track statistics
}
