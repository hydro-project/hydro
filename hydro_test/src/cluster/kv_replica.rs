use std::collections::HashMap;
use std::fmt::Debug;
use std::hash::Hash;

use hydro_lang::*;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};

pub struct Replica {}

pub trait KvKey: Serialize + DeserializeOwned + Hash + Eq + Clone + Debug {}
impl<K: Serialize + DeserializeOwned + Hash + Eq + Clone + Debug> KvKey for K {}

pub trait KvValue: Serialize + DeserializeOwned + Eq + Clone + Debug {}
impl<V: Serialize + DeserializeOwned + Eq + Clone + Debug> KvValue for V {}

#[derive(Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
pub struct KvPayload<K, V> {
    pub key: K,
    pub value: V,
}

#[derive(Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
pub struct SequencedKv<K, V> {
    // Note: Important that seq is the first member of the struct for sorting
    pub seq: usize,
    pub kv: Option<KvPayload<K, V>>,
}

impl<K: KvKey, V: KvValue> Ord for SequencedKv<K, V> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.seq.cmp(&other.seq)
    }
}

impl<K: KvKey, V: KvValue> PartialOrd for SequencedKv<K, V> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

// Replicas. All relations for replicas will be prefixed with r. Expects ReplicaPayload on p_to_replicas, outputs a stream of (client address, ReplicaPayload) after processing.
#[expect(clippy::type_complexity, reason = "internal paxos code // TODO")]
pub fn kv_replica<'a, K: KvKey, V: KvValue>(
    replicas: &Cluster<'a, Replica>,
    p_to_replicas: impl Into<
        Stream<(usize, Option<KvPayload<K, V>>), Cluster<'a, Replica>, Unbounded, NoOrder>,
    >,
    checkpoint_frequency: usize,
) -> (
    Stream<usize, Cluster<'a, Replica>, Unbounded>,
    Stream<KvPayload<K, V>, Cluster<'a, Replica>, Unbounded>,
) {
    let p_to_replicas: Stream<SequencedKv<K, V>, Cluster<'a, Replica>, Unbounded, NoOrder> =
        p_to_replicas
            .into()
            .map(q!(|(slot, kv)| SequencedKv { seq: slot, kv }));

    let replica_tick = replicas.tick();

    let (r_buffered_payloads_complete_cycle, r_buffered_payloads) = replica_tick.cycle();
    // p_to_replicas.inspect(q!(|payload: ReplicaPayload| println!("Replica received payload: {:?}", payload)));
    let r_sorted_payloads = unsafe {
        // SAFETY: because we fill slots one-by-one, we can safely batch
        // because non-determinism is resolved when we sort by slots
        p_to_replicas.tick_batch(&replica_tick)
    }
        .chain(r_buffered_payloads) // Combine with all payloads that we've received and not processed yet
        .sort();
    // Create a cycle since we'll use this seq before we define it
    let (r_next_slot_complete_cycle, r_next_slot) =
        replica_tick.cycle_with_initial(replica_tick.singleton(q!(0)));
    // Find highest the sequence number of any payload that can be processed in this tick. This is the payload right before a hole.
    let r_next_slot_after_processing_payloads = r_sorted_payloads
        .clone()
        .cross_singleton(r_next_slot.clone())
        .fold(
            q!(|| 0),
            q!(|new_next_slot, (sorted_payload, next_slot)| {
                if sorted_payload.seq == std::cmp::max(*new_next_slot, next_slot) {
                    *new_next_slot = sorted_payload.seq + 1;
                }
            }),
        );
    // Find all payloads that can and cannot be processed in this tick.
    let r_processable_payloads = r_sorted_payloads
        .clone()
        .cross_singleton(r_next_slot_after_processing_payloads.clone())
        .filter(q!(
            |(sorted_payload, highest_seq)| sorted_payload.seq < *highest_seq
        ))
        .map(q!(|(sorted_payload, _)| { sorted_payload }));
    let r_new_non_processable_payloads = r_sorted_payloads
        .clone()
        .cross_singleton(r_next_slot_after_processing_payloads.clone())
        .filter(q!(
            |(sorted_payload, highest_seq)| sorted_payload.seq > *highest_seq
        ))
        .map(q!(|(sorted_payload, _)| { sorted_payload }));
    // Save these, we can process them once the hole has been filled
    r_buffered_payloads_complete_cycle.complete_next_tick(r_new_non_processable_payloads);

    let r_kv_store = r_processable_payloads
        .clone()
        .persist() // Optimization: all_ticks() + fold() = fold<static>, where the state of the previous fold is saved and persisted values are deleted.
        .fold(q!(|| (HashMap::new(), 0)), q!(|(kv_store, next_slot), payload| {
            if let Some(kv) = payload.kv {
                kv_store.insert(kv.key, kv.value);
            }
            *next_slot = payload.seq + 1;
        }));
    // Update the highest seq for the next tick
    r_next_slot_complete_cycle
        .complete_next_tick(r_kv_store.map(q!(|(_kv_store, next_slot)| next_slot)));

    // Send checkpoints to the acceptors when we've processed enough payloads
    let (r_checkpointed_seqs_complete_cycle, r_checkpointed_seqs) =
        replica_tick.cycle::<Optional<usize, _, _>>();
    let r_max_checkpointed_seq = r_checkpointed_seqs.persist().max().into_singleton();
    let r_checkpoint_seq_new = r_max_checkpointed_seq
        .zip(r_next_slot)
        .filter_map(q!(
            move |(max_checkpointed_seq, next_slot)| if max_checkpointed_seq
                .map(|m| next_slot - m >= checkpoint_frequency)
                .unwrap_or(true)
            {
                Some(next_slot)
            } else {
                None
            }
        ));
    r_checkpointed_seqs_complete_cycle.complete_next_tick(r_checkpoint_seq_new.clone());

    // Tell clients that the payload has been committed. All ReplicaPayloads contain the client's machine ID (to string) as value.
    let r_to_clients = r_processable_payloads
        .filter_map(q!(|payload| payload.kv))
        .all_ticks();
    (r_checkpoint_seq_new.all_ticks(), r_to_clients)
}
