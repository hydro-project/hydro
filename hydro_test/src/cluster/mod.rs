pub mod compartmentalized_paxos;
pub mod compute_pi;
pub mod kv_replica;
pub mod many_to_many;
pub mod map_reduce;
pub mod paxos;
pub mod paxos_bench;
pub mod paxos_with_client;
pub mod simple_cluster;
pub mod two_pc;
