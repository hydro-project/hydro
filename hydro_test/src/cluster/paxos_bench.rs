use hydro_lang::*;
use hydro_std::quorum::collect_quorum;

use super::bench_client::{Client, bench_client};
use super::kv_replica::{KvPayload, Replica, kv_replica};
use super::paxos_with_client::PaxosLike;

pub fn paxos_bench<'a, Paxos: PaxosLike<'a>>(
    flow: &FlowBuilder<'a>,
    num_clients_per_node: usize,
    median_latency_window_size: usize, /* How many latencies to keep in the window for calculating the median */
    checkpoint_frequency: usize,       // How many sequence numbers to commit before checkpointing
    f: usize, /* Maximum number of faulty nodes. A payload has been processed once f+1 replicas have processed it. */
    num_replicas: usize,
    create_paxos: impl FnOnce(Stream<usize, Cluster<'a, Replica>, Unbounded>) -> Paxos,
) -> (Cluster<'a, Client>, Cluster<'a, Replica>) {
    let clients = flow.cluster::<Client>();
    let replicas = flow.cluster::<Replica>();

    bench_client(
        &clients,
        |c_to_proposers| {
            let payloads = c_to_proposers.map(q!(move |(key, value)| KvPayload {
                key,
                // we use our ID as part of the value and use that so the replica only notifies us
                value: (CLUSTER_SELF_ID, value)
            }));

            let (replica_checkpoint_complete, replica_checkpoint) =
                replicas.forward_ref::<Stream<_, _, _>>();

            let paxos = create_paxos(replica_checkpoint);

            let sequenced_payloads = unsafe {
                // SAFETY: clients "own" certain keys, so interleaving elements from clients will not affect
                // the order of writes to the same key

                // TODO(shadaj): we should retry when a payload is dropped due to stale leader
                paxos.with_client(&clients, payloads)
            };

            let sequenced_to_replicas = sequenced_payloads.broadcast_bincode_anonymous(&replicas);

            // Replicas
            let (replica_checkpoint, processed_payloads) =
                kv_replica(&replicas, sequenced_to_replicas, checkpoint_frequency);

            replica_checkpoint_complete.complete(replica_checkpoint);

            let c_received_payloads = processed_payloads
                .map(q!(|payload| (
                    payload.value.0,
                    ((payload.key, payload.value.1), Ok(()))
                )))
                .send_bincode_anonymous(&clients);

            // we only mark a transaction as committed when all replicas have applied it
            collect_quorum::<_, _, _, ()>(
                c_received_payloads.atomic(&clients.tick()),
                f + 1,
                num_replicas,
            )
            .0
            .end_atomic()
        },
        num_clients_per_node,
        median_latency_window_size,
    );

    (clients, replicas)
}

#[cfg(test)]
mod tests {
    use hydro_lang::deploy::DeployRuntime;
    use stageleft::RuntimeData;

    use crate::cluster::paxos::{CorePaxos, PaxosConfig};

    #[test]
    fn paxos_ir() {
        let builder = hydro_lang::FlowBuilder::new();
        let proposers = builder.cluster();
        let acceptors = builder.cluster();

        let _ = super::paxos_bench(&builder, 1, 1, 1, 1, 2, |replica_checkpoint| CorePaxos {
            proposers,
            acceptors: acceptors.clone(),
            replica_checkpoint: replica_checkpoint.broadcast_bincode(&acceptors),
            paxos_config: PaxosConfig {
                f: 1,
                i_am_leader_send_timeout: 1,
                i_am_leader_check_timeout: 1,
                i_am_leader_check_timeout_delay_multiplier: 1,
            },
        });
        let built = builder.with_default_optimize::<DeployRuntime>();

        hydro_lang::ir::dbg_dedup_tee(|| {
            insta::assert_debug_snapshot!(built.ir());
        });

        let _ = built.compile(&RuntimeData::new("FAKE"));
    }
}
