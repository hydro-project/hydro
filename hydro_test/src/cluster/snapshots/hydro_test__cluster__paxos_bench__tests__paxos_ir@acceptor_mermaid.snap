---
source: hydro_test/src/cluster/paxos_bench.rs
expression: "preview.dfir_for(&acceptors).to_mermaid(&WriteConfig\n{\n    no_subgraphs: true, no_pull_push: true, no_handoffs: true,\n    op_text_no_imports: true, ..WriteConfig::default()\n})"
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
classDef otherClass fill:#fdc,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa
1v1["<div style=text-align:center>(1v1)</div> <code><br>source_iter({<br>    [&quot;Acceptors say hello&quot;]<br>})</code>"]:::otherClass
2v1["<div style=text-align:center>(2v1)</div> <code><br>for_each({<br>    |s| println!(&quot;{}&quot;, s)<br>})</code>"]:::otherClass
3v1["<div style=text-align:center>(3v1)</div> <code><br>source_stream(DUMMY_SOURCE)</code>"]:::otherClass
4v1["<div style=text-align:center>(4v1)</div> <code><br>map(|res| {<br>    let (id, b) = res.unwrap();<br>    (<br>        hydro_lang::ClusterId::&lt;<br>            hydro_test::__staged::cluster::paxos::Proposer,<br>        &gt;::from_raw(id),<br>        hydro_lang::runtime_support::bincode::deserialize::&lt;<br>            hydro_test::__staged::cluster::paxos::Ballot,<br>        &gt;(&amp;b)<br>            .unwrap(),<br>    )<br>})</code>"]:::otherClass
5v1["<div style=text-align:center>(5v1)</div> <code><br>map({<br>    |(_, b)| b<br>})</code>"]:::otherClass
6v1["<div style=text-align:center>(6v1)</div> <code><br>tee()</code>"]:::otherClass
7v1["<div style=text-align:center>(7v1)</div> <code><br>inspect({<br>    |p1a| println!(&quot;Acceptor received P1a: {:?}&quot;, p1a)<br>})</code>"]:::otherClass
8v1["<div style=text-align:center>(8v1)</div> <code><br>reduce::&lt;<br>    'static,<br>&gt;({<br>    |curr, new| {<br>        if new &gt; *curr {<br>            *curr = new;<br>        }<br>    }<br>})</code>"]:::otherClass
9v1["<div style=text-align:center>(9v1)</div> <code><br>source_iter({<br>    let e__free = {<br>        Ballot {<br>            num: 0,<br>            proposer_id: ClusterId::from_raw(0),<br>        }<br>    };<br>    [e__free]<br>})</code>"]:::otherClass
10v1["<div style=text-align:center>(10v1)</div> <code><br>persist::&lt;'static&gt;()</code>"]:::otherClass
11v1["<div style=text-align:center>(11v1)</div> <code><br>chain()</code>"]:::otherClass
12v1["<div style=text-align:center>(12v1)</div> <code><br>tee()</code>"]:::otherClass
13v1["<div style=text-align:center>(13v1)</div> <code><br>cross_singleton()</code>"]:::otherClass
14v1["<div style=text-align:center>(14v1)</div> <code><br>cross_singleton()</code>"]:::otherClass
15v1["<div style=text-align:center>(15v1)</div> <code><br>map({<br>    |((ballot, max_ballot), log)| (<br>        ballot.proposer_id,<br>        (ballot, if ballot == max_ballot { Ok(log) } else { Err(max_ballot) }),<br>    )<br>})</code>"]:::otherClass
16v1["<div style=text-align:center>(16v1)</div> <code><br>map(|(id, data)| {<br>    (<br>        id.raw_id,<br>        hydro_lang::runtime_support::bincode::serialize(&amp;data).unwrap().into(),<br>    )<br>})</code>"]:::otherClass
17v1["<div style=text-align:center>(17v1)</div> <code><br>dest_sink(DUMMY_SINK)</code>"]:::otherClass
18v1["<div style=text-align:center>(18v1)</div> <code><br>source_stream(DUMMY_SOURCE)</code>"]:::otherClass
19v1["<div style=text-align:center>(19v1)</div> <code><br>map(|res| {<br>    let (id, b) = res.unwrap();<br>    (<br>        hydro_lang::ClusterId::&lt;<br>            hydro_test::__staged::cluster::paxos::Proposer,<br>        &gt;::from_raw(id),<br>        hydro_lang::runtime_support::bincode::deserialize::&lt;<br>            hydro_test::__staged::cluster::paxos::P2a&lt;<br>                hydro_test::__staged::cluster::kv_replica::KvPayload&lt;<br>                    u32,<br>                    (<br>                        hydro_test::__staged::__deps::hydro_lang::location::cluster::cluster_id::ClusterId&lt;<br>                            hydro_test::__staged::cluster::paxos_bench::Client,<br>                        &gt;,<br>                        u32,<br>                    ),<br>                &gt;,<br>                hydro_test::__staged::cluster::paxos::Proposer,<br>            &gt;,<br>        &gt;(&amp;b)<br>            .unwrap(),<br>    )<br>})</code>"]:::otherClass
20v1["<div style=text-align:center>(20v1)</div> <code><br>map({<br>    |(_, b)| b<br>})</code>"]:::otherClass
21v1["<div style=text-align:center>(21v1)</div> <code><br>tee()</code>"]:::otherClass
22v1["<div style=text-align:center>(22v1)</div> <code><br>cross_singleton()</code>"]:::otherClass
23v1["<div style=text-align:center>(23v1)</div> <code><br>map({<br>    |(p2a, max_ballot)| (<br>        p2a.sender,<br>        (<br>            (p2a.slot, p2a.ballot),<br>            if p2a.ballot == max_ballot { Ok(()) } else { Err(max_ballot) },<br>        ),<br>    )<br>})</code>"]:::otherClass
24v1["<div style=text-align:center>(24v1)</div> <code><br>map(|(id, data)| {<br>    (<br>        id.raw_id,<br>        hydro_lang::runtime_support::bincode::serialize(&amp;data).unwrap().into(),<br>    )<br>})</code>"]:::otherClass
25v1["<div style=text-align:center>(25v1)</div> <code><br>dest_sink(DUMMY_SINK)</code>"]:::otherClass
26v1["<div style=text-align:center>(26v1)</div> <code><br>cross_singleton()</code>"]:::otherClass
27v1["<div style=text-align:center>(27v1)</div> <code><br>filter_map({<br>    |(p2a, max_ballot)| {<br>        if p2a.ballot &gt;= max_ballot { Some(CheckpointOrP2a::P2a(p2a)) } else { None }<br>    }<br>})</code>"]:::otherClass
28v1["<div style=text-align:center>(28v1)</div> <code><br>multiset_delta()</code>"]:::otherClass
29v1["<div style=text-align:center>(29v1)</div> <code><br>map({<br>    |min_seq| CheckpointOrP2a::Checkpoint(min_seq)<br>})</code>"]:::otherClass
30v1["<div style=text-align:center>(30v1)</div> <code><br>chain()</code>"]:::otherClass
31v1["<div style=text-align:center>(31v1)</div> <code><br>fold::&lt;<br>    'static,<br>&gt;(<br>    {<br>        || (None, HashMap::new())<br>    },<br>    {<br>        |(prev_checkpoint, log), checkpoint_or_p2a| {<br>            match checkpoint_or_p2a {<br>                CheckpointOrP2a::Checkpoint(new_checkpoint) =&gt; {<br>                    if prev_checkpoint<br>                        .map(|prev| new_checkpoint &gt; prev)<br>                        .unwrap_or(true)<br>                    {<br>                        for slot in (prev_checkpoint.unwrap_or(0))..new_checkpoint {<br>                            log.remove(&amp;slot);<br>                        }<br>                        *prev_checkpoint = Some(new_checkpoint);<br>                    }<br>                }<br>                CheckpointOrP2a::P2a(p2a) =&gt; {<br>                    if prev_checkpoint.map(|prev| p2a.slot &gt; prev).unwrap_or(true)<br>                        &amp;&amp; log<br>                            .get(&amp;p2a.slot)<br>                            .map(|prev_p2a: &amp;LogValue&lt;_&gt;| p2a.ballot &gt; prev_p2a.ballot)<br>                            .unwrap_or(true)<br>                    {<br>                        log.insert(<br>                            p2a.slot,<br>                            LogValue {<br>                                ballot: p2a.ballot,<br>                                value: p2a.value,<br>                            },<br>                        );<br>                    }<br>                }<br>            }<br>        }<br>    },<br>)</code>"]:::otherClass
32v1["<div style=text-align:center>(32v1)</div> <code><br>source_stream(DUMMY_SOURCE)</code>"]:::otherClass
33v1["<div style=text-align:center>(33v1)</div> <code><br>map(|res| {<br>    let (id, b) = res.unwrap();<br>    (<br>        hydro_lang::ClusterId::&lt;<br>            hydro_test::__staged::cluster::kv_replica::Replica,<br>        &gt;::from_raw(id),<br>        hydro_lang::runtime_support::bincode::deserialize::&lt;usize&gt;(&amp;b).unwrap(),<br>    )<br>})</code>"]:::otherClass
34v1["<div style=text-align:center>(34v1)</div> <code><br>reduce_keyed::&lt;<br>    'static,<br>&gt;({<br>    |curr_seq, seq| {<br>        if seq &gt; *curr_seq {<br>            *curr_seq = seq;<br>        }<br>    }<br>})</code>"]:::otherClass
35v1["<div style=text-align:center>(35v1)</div> <code><br>tee()</code>"]:::otherClass
36v1["<div style=text-align:center>(36v1)</div> <code><br>fold::&lt;<br>    'tick,<br>&gt;(<br>    {<br>        || 0usize<br>    },<br>    {<br>        |count, _| *count += 1<br>    },<br>)</code>"]:::otherClass
37v1["<div style=text-align:center>(37v1)</div> <code><br>filter_map({<br>    let f__free = 1usize;<br>    move |num_received| if num_received == f__free + 1 { Some(true) } else { None }<br>})</code>"]:::otherClass
38v1["<div style=text-align:center>(38v1)</div> <code><br>map({<br>    |_u| ()<br>})</code>"]:::otherClass
39v1["<div style=text-align:center>(39v1)</div> <code><br>cross_singleton()</code>"]:::otherClass
40v1["<div style=text-align:center>(40v1)</div> <code><br>map({<br>    |(d, _signal)| d<br>})</code>"]:::otherClass
41v1["<div style=text-align:center>(41v1)</div> <code><br>map({<br>    |(_sender, seq)| seq<br>})</code>"]:::otherClass
42v1["<div style=text-align:center>(42v1)</div> <code><br>reduce::&lt;<br>    'tick,<br>&gt;({<br>    |curr, new| {<br>        if new &lt; *curr {<br>            *curr = new;<br>        }<br>    }<br>})</code>"]:::otherClass
1v1-->2v1
3v1-->4v1
4v1-->5v1
5v1-->6v1
6v1-->7v1
7v1--x8v1; linkStyle 5 stroke:red
9v1-->10v1
8v1--x|0|11v1; linkStyle 7 stroke:red
10v1-->|1|11v1
11v1-->12v1
6v1-->|input|13v1
12v1--x|single|13v1; linkStyle 11 stroke:red
13v1-->|input|14v1
31v1--x|single|14v1; linkStyle 13 stroke:red
14v1-->15v1
16v1-->17v1
15v1-->16v1
18v1-->19v1
19v1-->20v1
20v1-->21v1
21v1-->|input|22v1
12v1--x|single|22v1; linkStyle 21 stroke:red
22v1-->23v1
24v1-->25v1
23v1-->24v1
21v1-->|input|26v1
12v1--x|single|26v1; linkStyle 26 stroke:red
26v1-->27v1
42v1-->28v1
28v1-->29v1
27v1--x|0|30v1; linkStyle 30 stroke:red
29v1-->|1|30v1
30v1--x31v1; linkStyle 32 stroke:red
32v1-->33v1
33v1--x34v1; linkStyle 34 stroke:red
34v1-->35v1
35v1--x36v1; linkStyle 36 stroke:red
36v1-->37v1
37v1-->38v1
35v1-->|input|39v1
38v1--x|single|39v1; linkStyle 40 stroke:red
39v1-->40v1
40v1-->41v1
41v1--x42v1; linkStyle 43 stroke:red
2v1
16v1
17v1
24v1
25v1
subgraph var_stream_2 ["var <tt>stream_2</tt>"]
    style var_stream_2 fill:transparent
    1v1
end
subgraph var_stream_237 ["var <tt>stream_237</tt>"]
    style var_stream_237 fill:transparent
    18v1
    19v1
end
subgraph var_stream_238 ["var <tt>stream_238</tt>"]
    style var_stream_238 fill:transparent
    20v1
end
subgraph var_stream_239 ["var <tt>stream_239</tt>"]
    style var_stream_239 fill:transparent
    21v1
end
subgraph var_stream_241 ["var <tt>stream_241</tt>"]
    style var_stream_241 fill:transparent
    22v1
end
subgraph var_stream_242 ["var <tt>stream_242</tt>"]
    style var_stream_242 fill:transparent
    23v1
end
subgraph var_stream_275 ["var <tt>stream_275</tt>"]
    style var_stream_275 fill:transparent
    26v1
end
subgraph var_stream_276 ["var <tt>stream_276</tt>"]
    style var_stream_276 fill:transparent
    27v1
end
subgraph var_stream_278 ["var <tt>stream_278</tt>"]
    style var_stream_278 fill:transparent
    28v1
end
subgraph var_stream_279 ["var <tt>stream_279</tt>"]
    style var_stream_279 fill:transparent
    29v1
end
subgraph var_stream_280 ["var <tt>stream_280</tt>"]
    style var_stream_280 fill:transparent
    30v1
end
subgraph var_stream_281 ["var <tt>stream_281</tt>"]
    style var_stream_281 fill:transparent
    31v1
end
subgraph var_stream_339 ["var <tt>stream_339</tt>"]
    style var_stream_339 fill:transparent
    32v1
    33v1
end
subgraph var_stream_340 ["var <tt>stream_340</tt>"]
    style var_stream_340 fill:transparent
    34v1
end
subgraph var_stream_341 ["var <tt>stream_341</tt>"]
    style var_stream_341 fill:transparent
    35v1
end
subgraph var_stream_343 ["var <tt>stream_343</tt>"]
    style var_stream_343 fill:transparent
    36v1
end
subgraph var_stream_344 ["var <tt>stream_344</tt>"]
    style var_stream_344 fill:transparent
    37v1
end
subgraph var_stream_345 ["var <tt>stream_345</tt>"]
    style var_stream_345 fill:transparent
    38v1
end
subgraph var_stream_346 ["var <tt>stream_346</tt>"]
    style var_stream_346 fill:transparent
    39v1
end
subgraph var_stream_347 ["var <tt>stream_347</tt>"]
    style var_stream_347 fill:transparent
    40v1
end
subgraph var_stream_348 ["var <tt>stream_348</tt>"]
    style var_stream_348 fill:transparent
    41v1
end
subgraph var_stream_349 ["var <tt>stream_349</tt>"]
    style var_stream_349 fill:transparent
    42v1
end
subgraph var_stream_66 ["var <tt>stream_66</tt>"]
    style var_stream_66 fill:transparent
    3v1
    4v1
end
subgraph var_stream_67 ["var <tt>stream_67</tt>"]
    style var_stream_67 fill:transparent
    5v1
end
subgraph var_stream_68 ["var <tt>stream_68</tt>"]
    style var_stream_68 fill:transparent
    6v1
end
subgraph var_stream_70 ["var <tt>stream_70</tt>"]
    style var_stream_70 fill:transparent
    7v1
end
subgraph var_stream_71 ["var <tt>stream_71</tt>"]
    style var_stream_71 fill:transparent
    8v1
end
subgraph var_stream_72 ["var <tt>stream_72</tt>"]
    style var_stream_72 fill:transparent
    9v1
end
subgraph var_stream_73 ["var <tt>stream_73</tt>"]
    style var_stream_73 fill:transparent
    10v1
end
subgraph var_stream_74 ["var <tt>stream_74</tt>"]
    style var_stream_74 fill:transparent
    11v1
end
subgraph var_stream_75 ["var <tt>stream_75</tt>"]
    style var_stream_75 fill:transparent
    12v1
end
subgraph var_stream_76 ["var <tt>stream_76</tt>"]
    style var_stream_76 fill:transparent
    13v1
end
subgraph var_stream_78 ["var <tt>stream_78</tt>"]
    style var_stream_78 fill:transparent
    14v1
end
subgraph var_stream_79 ["var <tt>stream_79</tt>"]
    style var_stream_79 fill:transparent
    15v1
end
